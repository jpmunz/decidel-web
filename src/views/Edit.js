import React from "react";
import { Container, Row, Col } from "react-bootstrap";
import { connect } from "react-redux";
import {
  fetchCurrentUserRequest,
  fetchDecidelRequest,
  getFinalDecision,
  getAsyncRequest,
  hasDecidel,
  getTitle
} from "src/state";
import DecisionOptions from "./DecisionOptions";
import Deciders from "./Deciders";
import History from "./History";
import Loading from "./Loading";
import CurrentRound from "./CurrentRound";
import "./Edit.scss";

export class Edit extends React.Component {
  componentDidMount() {
    this.props.fetchDecidelRequest({ id: this.props.id });
    this.props.fetchCurrentUserRequest();
  }

  render() {
    let content;

    if (this.props.request.error) {
      content = (
        <>
          <Row className="justify-content-center">
            <Col className="decidel-not-found">
              <div>We couldn't find a Decidel with that id.</div>
              <div>Are you sure you entered the URL correctly?</div>
            </Col>
          </Row>
        </>
      );
    } else if (!this.props.decidelLoaded || this.props.request.inProgress) {
      return <Loading />;
    } else if (this.props.finalDecision) {
      content = (
        <>
          <Row>
            <Col>
              <div className={"final-decision-header"}>
                <div className={"final-decision-preamble"}>You decided on</div>
                <div className={"final-decision-title"}>
                  {this.props.finalDecision + "!"}
                </div>
              </div>
            </Col>
          </Row>
          <Row className="justify-content-center">
            <Col className={"col-8"}>
              <DecisionOptions id={this.props.id} />
            </Col>
          </Row>
          <Row className="justify-content-center">
            <Col className={"col-8"}>
              <History id={this.props.id} />
            </Col>
          </Row>
        </>
      );
    } else {
      content = (
        <>
          <Row>
            <Col>
              <h1>{this.props.title}</h1>
            </Col>
          </Row>
          <Row>
            <Col>
              <Row>
                <Col>
                  <CurrentRound id={this.props.id} />
                </Col>
              </Row>
              <Row>
                <Col>
                  <DecisionOptions id={this.props.id} />
                </Col>
              </Row>
            </Col>
            <Col className="col-sm">
              <Deciders id={this.props.id} />
            </Col>
          </Row>
          <Row>
            <Col>
              <History id={this.props.id} />
            </Col>
          </Row>
        </>
      );
    }

    return <Container className="Edit">{content}</Container>;
  }
}

export default connect(
  (state, props) => {
    return {
      finalDecision: getFinalDecision(state, props.id),
      request: getAsyncRequest(state, props.id),
      title: getTitle(state, props.id),
      decidelLoaded: hasDecidel(state, props.id)
    };
  },
  { fetchDecidelRequest, fetchCurrentUserRequest }
)(Edit);
