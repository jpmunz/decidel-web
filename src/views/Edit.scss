@import "mixins";

.Edit {
  h1 {
    padding-bottom: 20px;
    text-align: center;
  }

  .History {
    padding-top: 30px;
  }

  .spinner {
    text-align: center;
  }

  .final-decision-header {
    text-align: center;
    margin-bottom: 20px;

    .final-decision-preamble {
      font-size: 3em;
    }

    .final-decision-title {
      position: relative;
      top: -20px;
      font-weight: bold;
      font-size: 3.5em;
    }
  }

  .decidel-not-found {
    text-align: center;
    font-size: 1.5em;
  }
}
