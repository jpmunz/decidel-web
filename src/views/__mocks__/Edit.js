import React from "react";

export const MockEdit = jest.fn();

MockEdit.mockReturnValue(<div>Mocked Edit</div>);

const mock = jest.fn().mockImplementation(MockEdit);

export default mock;
