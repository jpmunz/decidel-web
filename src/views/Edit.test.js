import React from "react";
import { Edit } from "./Edit";
import { render } from "@testing-library/react";

jest.mock("./History");
jest.mock("./DecisionOptions");
jest.mock("./CurrentRound");

jest.mock("./Deciders");

const fetchDecidelMock = jest.fn();
const fetchCurrentUserMock = jest.fn();

const renderComponent = function({
  id = "22",
  title = "What movie to watch",
  finalDecision = false,
  request = {},
  currentUserName = "Bob",
  decidelLoaded = true
}) {
  return render(
    <Edit
      id={id}
      title={title}
      finalDecision={finalDecision}
      currentUserName={currentUserName}
      decidelLoaded={decidelLoaded}
      request={request}
      fetchDecidelRequest={fetchDecidelMock}
      fetchCurrentUserRequest={fetchCurrentUserMock}
    />
  );
};

const renderTest = options =>
  expect(renderComponent(options).asFragment()).toMatchSnapshot();

test("Initial mount", () => renderTest({ decidelLoaded: false }));
test("Fetching in progress", () =>
  renderTest({ request: { inProgress: true } }));
test("Decision finalized", () => renderTest({ finalDecision: "The Rock" }));
test("Fetch failed", () => renderTest({ request: { error: "failed" } }));

test("Fetches are made on mount", () => {
  renderComponent({});

  expect(fetchDecidelMock.mock.calls).toEqual([[{ id: "22" }]]);
  expect(fetchCurrentUserMock.mock.calls).toEqual([[]]);
});
