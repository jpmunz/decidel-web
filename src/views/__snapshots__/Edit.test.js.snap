// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Decision finalized 1`] = `
<DocumentFragment>
  <div
    class="Edit container"
  >
    <div
      class="row"
    >
      <div
        class="col"
      >
        <div
          class="final-decision-header"
        >
          <div
            class="final-decision-preamble"
          >
            You decided on
          </div>
          <div
            class="final-decision-title"
          >
            The Rock!
          </div>
        </div>
      </div>
    </div>
    <div
      class="justify-content-center row"
    >
      <div
        class="col-8 col"
      >
        <div>
          Mocked DecisionOptions
        </div>
      </div>
    </div>
    <div
      class="justify-content-center row"
    >
      <div
        class="col-8 col"
      >
        <div>
          Mocked History
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Fetch failed 1`] = `
<DocumentFragment>
  <div
    class="Edit container"
  >
    <div
      class="justify-content-center row"
    >
      <div
        class="decidel-not-found col"
      >
        <div>
          We couldn't find a Decidel with that id.
        </div>
        <div>
          Are you sure you entered the URL correctly?
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Fetching in progress 1`] = `
<DocumentFragment>
  <div
    class="Loading"
  >
    <div
      class="spinner-border text-primary"
      role="status"
    >
      <span
        class="sr-only"
      >
        Loading...
      </span>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Initial mount 1`] = `
<DocumentFragment>
  <div
    class="Loading"
  >
    <div
      class="spinner-border text-primary"
      role="status"
    >
      <span
        class="sr-only"
      >
        Loading...
      </span>
    </div>
  </div>
</DocumentFragment>
`;
